kernel/src/net/iface/init.rs

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{borrow::ToOwned, sync::Arc};
+
+use aster_bigtcp::device::WithDevice;
+use ostd::sync::PreemptDisabled;
+use spin::Once;
+
+use super::{poll_ifaces, Iface};
+use crate::{
+    net::iface::ext::{IfaceEx, IfaceExt},
+    prelude::*,
+};
+
+pub static IFACES: Once<Vec<Arc<Iface>>> = Once::new();
+
+pub fn init() {
+    IFACES.call_once(|| {
+        let iface_virtio = new_virtio();
+        let iface_loopback = new_loopback();
+        vec![iface_virtio, iface_loopback]
+    });
+
+    for (name, _) in aster_network::all_devices() {
+        aster_network::register_recv_callback(&name, || {
+            // TODO: further check that the irq num is the same as iface's irq num
+            let iface_virtio = &IFACES.get().unwrap()[0];
+            iface_virtio.poll();
+        })
+    }
+
+    poll_ifaces();
+}
+
+fn new_virtio() -> Arc<Iface> {
+    use aster_bigtcp::{iface::EtherIface, wire::EthernetAddress};
+    use aster_network::AnyNetworkDevice;
+    use aster_virtio::device::network::DEVICE_NAME;
+
+    let virtio_net = aster_network::get_device(DEVICE_NAME).unwrap();
+
+    let ether_addr = virtio_net.disable_irq().lock().mac_addr().0;
+
+    struct Wrapper(Arc<SpinLock<dyn AnyNetworkDevice, PreemptDisabled>>);
+
```

Comment:
Why not just use `Arc<SpinLock<dyn AnyNetworkDevice, LocalIrqDisabled>>)`? It seems you will always disable irq when visiting this device?

Reply:
This is for historical reasons. I've added a new commit to take advantage of our new `SpinLock` APIs.